Parse env-deps from dep-info

jschwe opened the pull request:
Parse the environment variables (and their values) referenced in rust code via `env!` or `option_env!` from the `dep-info` file and add them to the hash. The `CARGO_` env variables apparently still have to be blanket added to the hash, since they don't appear in the `dep-info` file.

Closes #972 
Closes #804

mitchhentges reviewed (changes requested):
Looks pretty good, though I'm mostly just marking this as requesting changes for the extra log lines.

Thanks for the patch :)

jschwe commented on src/compiler/rust.rs:
```diff
@@ -316,6 +317,27 @@ where
     deps
 }
 
+fn parse_env_dep_info(dep_info: &str) -> Vec<(OsString, OsString)> {
```
Actually, the first pass only reads the first line of the `dep_info` file, so I think from the performance perspective there is no benefit in doing so. Keeping it separate helps readability in my opinion.

jschwe commented on tests/sccache_cargo.rs:
```diff
@@ -133,5 +142,34 @@ fn test_rust_cargo_cmd(cmd: &str, extra_envs: &[(&str, &std::ffi::OsStr)]) {
         .stdout(predicates::str::contains(r#""cache_hits":{"counts":{"Rust":2}}"#).from_utf8())
         .success();
+
+    if cmd == "run" {
```
The test for the env vars is now separate. While doing so I refactored basically the whole test and I hope it is easier to add new tests now without duplicating a lot of code.

I removed the deny for rust 2018 idioms from the integration test file, since the test itself should always be built with a cargo version >= the sccache MSRV (which I guessed to be 1.48 by looking at CI). 

Depending on how important the `cfg` excluding windows and macos was, I may also just put everything in the test into a mod.

jschwe commented:
This might need another round of iterations, but if everything is okay, please ping me so I can rebase and squash before merging.

mitchhentges commented on tests/sccache_cargo.rs:
```diff
@@ -64,74 +112,123 @@ fn test_rust_cargo_cmd(cmd: &str, extra_envs: &[(&str, &std::ffi::OsStr)]) {
+        .try_stdout(predicates::str::contains(r#""cache_hits":{"counts":{"Rust":2}}"#).from_utf8())?
+        .try_success()?;
+
+    drop(test_info);
```
Isn't `test_info` dropped automatically at the end of this function?
If we need to explicitly stop the server to make it shut down between tests, perhaps we should do so explicitly without leaning on `drop()`?

jschwe replied:
> Isn't test_info dropped automatically at the end of this function?

Yes. But I was/am not sure if it might even be dropped _before_ the end of the function. I was thinking that by explicitly dropping it at the end, I could prevent it from being dropped after its last usage, since the server still has to be alive for the "get stats" step. I guess I probably should make the "show stats" invocation an associated function of `test_info`, that way the server would be guaranteed to still be alive.

mitchhentges commented on tests/sccache_cargo.rs:
```diff
@@ -3,53 +3,101 @@
-    test_rust_cargo_cmd("check", &[("RUSTFLAGS", std::ffi::OsStr::new("-Zprofile"))]);
-    #[cfg(feature = "unstable")]
-    test_rust_cargo_cmd("build", &[("RUSTFLAGS", std::ffi::OsStr::new("-Zprofile"))]);
+/// Used as a test setup fixture. The drop implementation cleans up after a _successfull_ test.
```
```suggestion
/// Used as a test setup fixture. The drop implementation cleans up after a _successful_ test.
```

mitchhentges commented on tests/sccache_cargo.rs:
```diff
@@ -3,53 +3,101 @@
-    #[cfg(feature = "unstable")]
-    test_rust_cargo_cmd("build", &[("RUSTFLAGS", std::ffi::OsStr::new("-Zprofile"))]);
+/// Used as a test setup fixture. The drop implementation cleans up after a _successfull_ test.
+/// It is therefore important that the asserts use `Result` to return errors instead of panicking.
+/// Otherwise the temporary directory will not be cleaned up and the started sccache server process
+/// will not be stopped.
```
If we need to do post-test cleanup, we can catch the unwind:
```
let test_result = panic::catch_unwind(|| {
    // do majority of the test
});
// do test cleanup
assert!(test_result.is_ok())
```

This will help avoid a footgun for consumers of `SccacheTest`.

jschwe replied:
For now I just immediately unwrap again after `catch_unwind`, since the `drop` for the `SccacheTest` would have already run then. The unwrapping returns the result of the test, which is maybe a bit unintuitive.

jschwe replied:
Now that I think it again, the `drop` actually runs anyway if `panic=unwind` and will clean up the directory. And if `panic=abort`, then the catch_unwind will also not work so I'm just going to remove it again, since it doesn't help.

mitchhentges commented on tests/sccache_cargo.rs:
```diff
@@ -3,53 +3,101 @@
+/// Otherwise the temporary directory will not be cleaned up and the started sccache server process
+/// will not be stopped.
+struct SccacheTest<'a> {
+    /// Tempdir used for Sccache cache and cargo output. It is kept in the struct only to have the
+    /// destructor run when SccacheTest goes out of scope, but is never used otherwise.
+    #[allow(dead_code)]
+    tempdir: tempfile::TempDir,
```
Perhaps have a `run_cargo()` method on `SccacheTest` that builds the `env` dict so that the compiler can more easily understand the lifecycle of the `tempdir`. As it is now, we still have "references" to the `tempdir` in the form of raw string paths, but this is too abstract for the compiler to track the lifetime automatically.

jschwe replied:
I added the `show_stats()` impl on the SccacheTest struct, so that the struct is used when running `sccache --show-stats`. 

I double checked, and [the reference](https://doc.rust-lang.org/reference/destructors.html#scopes-of-function-parameters) guarantees that function parameters are dropped last, at the very end of the function. Struct members also can't be dropped prematurely, so I would prefer to not complicate things and just keep it as is.

mitchhentges commented on tests/sccache_cargo.rs:
```diff
@@ -64,74 +112,123 @@ fn test_rust_cargo_cmd(cmd: &str, extra_envs: &[(&str, &std::ffi::OsStr)]) {
+    test_rust_cargo_cmd("build", SccacheTest::new(None)?)
+        .context("Sccache failed for `cargo build`")?;
+
+    test_rust_cargo_env_dep(SccacheTest::new(None)?)
+        .context("Sccache failed for `cargo build` with changed environment variables.")?;
```
Let's split these into three test functions:
* One checking that a base cargo build has cache hits
* One checking that a cargo `-Zprofile` build has cache hits
* One checking that changing a needed env var has a cache miss.

jschwe replied:
I actually tried that out, but there is a complication. Cargo test by default invokes multiple tests in parallel - and this doesn't work well with sccache. We already use different directories, but sccache complains "address already in use" when starting a second sccache server.

So you can only test sequentially. There is an option for cargo test to do that, but by default `cargo test` would result in failing tests. 
So we would have to add a bit of extra complexity - E.g. use a Mutex (but keep lock poisining in mind) or an atomic variable to synchronize access and make sure only one test runs at a time, even if multiple are started in parallel.

jschwe replied:
Thanks for the suggestion - `serial_test` worked just fine.

mitchhentges reviewed (changes requested):
Looks pretty good, almost there! I want to avoid `drop()` footguns here, then we should be good :)

jschwe commented on tests/sccache_cargo.rs:
```diff
@@ -3,53 +3,101 @@
-fn test_rust_cargo() {
-    test_rust_cargo_cmd("check", &[]);
-    test_rust_cargo_cmd("build", &[]);
+static SCCACHE_BIN: Lazy<PathBuf> = Lazy::new(|| assert_cmd::cargo::cargo_bin("sccache"));
+static CARGO: Lazy<OsString> = Lazy::new(|| std::env::var_os("CARGO").unwrap());
+static CRATE_DIR: Lazy<PathBuf> =
+    Lazy::new(|| Path::new(file!()).parent().unwrap().join("test-crate"));
```
The documentation for `assert_cmd::cargo::cargo_bin` recommends to cache the result for performance reason. I'm not quite sure how up to date this is, but the issue linked from the documentation mentioned something like up to 0.1-0.3 seconds per evaluation.

For the other ones I guess it is not so important and I could just replace the lazy static with a function.

mitchhentges replied:
> The documentation for assert_cmd::cargo::cargo_bin recommends to cache the result for performance reason.

Good dig, nice work.
FWIW, I'm assuming that that performance reason only matters if the variable is sometimes unused. Since we're guaranteed to need `SCCACHE_BIN` for this test, we're guaranteed to be running the `Lazy` closure, and therefore it can be inlined I believe.

jschwe replied:
> FWIW, I'm assuming that that performance reason only matters if the variable is sometimes unused.

No, this is not the reason. once_cell::Lazy, is initialized on first access (hence lazy), but on any further accesses the value is already available and the closure is not executed. So the performance benefit this is used for is if you are evaluating something multiple times and know that the result can be cached.

jschwe replied:
I guess it would be more clear if I used the full path for the type annotation, since only `Lazy` is somewhat ambiguous.

mitchhentges replied:
> No, this is not the reason. once_cell::Lazy, is initialized on first access (hence lazy), but on any further accesses the value is already available and the closure is not executed

Right, but without `Lazy`, it's equivalent to the closure work (in our case, the code `assert_cmd::cargo::cargo_bin("sccache")`) being run immediately at runtime. Either way, the value is re-used. By removing `Lazy`, the only functional change that occurs is that the "closure work" happens earlier, IIUC.

So, the only way that `Lazy` would be associated with a performance benefit is if, due to conditional logic, the variable is never accessed, so the closure work is never redundantly resolved.
Since we always access the variable in our tests, this performance benefit is never encountered, right?

jschwe replied:
> Right, but without Lazy, it's equivalent to the closure work being run immediately at runtime

I'm not sure I follow. Do you mean something like
```
 static CARGO: String = std::env::var("CARGO").unwrap(); 
```

That doesn't work because
```
error[E0015]: calls in statics are limited to constant functions, tuple structs and tuple variants
```

So I'm not sure if I'm missing something, but I don't see how to avoid using a lazily initialized variable for a global that can't be initialized at compile time.

mitchhentges replied:
Oh, I see, I was imagining something like:
```suggestion
const SCCACHE_BIN: PathBuf = assert_cmd::cargo::cargo_bin("sccache");
const CARGO: OsString = std::env::var_os("CARGO").unwrap();
const CRATE_DIR: PathBuf = Path::new(file!()).parent().unwrap().join("test-crate");
```

But, it looks like quite a few of these functions aren't `const`-able yet.
Ok, let's carry on with `Lazy`, my mistake :)

jschwe commented:
Just FYI, I haven't gotten to addressing the review comments yet. The force push just now didn't change the end result, I just rebased, squashed and reordered the commits.

jschwe commented:
Changes are in commits since  [fixup: Rename TEST_ENV_VAR](https://github.com/mozilla/sccache/pull/1107/commits/b0f162d96d565a8536cea7a6916a9752701d6363). I'd rebase and squash again after review.

jschwe commented on tests/sccache_cargo.rs:
```diff
@@ -37,6 +53,10 @@ struct SccacheTest<'a> {
 
 impl SccacheTest<'_> {
     fn new(additional_envs: Option<&[(&'static str, std::ffi::OsString)]>) -> Result<Self> {
+        assert!(LOGGER.is_ok());
```
This one is a bit tricky and maybe there is a better way to do that. 
We need to initialize the logger exactly once. Running `LOGGER.is_ok()` actually initializes the logger the first time. The `is_ok()` has to be used so I just put it into the assert.

mitchhentges commented:
Looks good, thanks!
